lib/internal/modules/esm/importmap.js

```diff
@@ -0,0 +1,135 @@
+    } catch (err) {
+      if (err.code !== 'ENOENT') {
+        setOwnProperty(err, 'importMapPath', `${filepath}`);
+        throw err;
+      }
+      return;
+    }
+
+    let json;
+    try {
+      json = JSONParse(str);
+    } catch (err) {
+      setOwnProperty(err, 'importMap', `${str}`);
+      throw err;
+    }
+
+    return new ImportMap(json, pathToFileURL(path.dirname(filepath)));
+  }
+
+  resolve(specifier, parentURL) {
+    const scope = parentURL || this.baseURL;
+
+    // Process scopes
+    // @TODO this almost certainly needs to be a radix tree
+    for (const [prefix, mapping] of this.scopes) {
+      let mappedSpecifier = mapping.get(specifier);
+      if (scope.pathname.startsWith(prefix.pathname) && mappedSpecifier) {
+        if (!isURL(mappedSpecifier)) {
+          mappedSpecifier = pathToFileURL(mappedSpecifier, this.baseURL.pathname);
+          mapping.set(specifier, mappedSpecifier);
+        }
+        specifier = mappedSpecifier;
+        break;
+      }
+    }
+
+    let spec = specifier;
+    if (isURL(specifier)) {
+      spec = specifier.pathname;
+    }
+    let importMapping = this.imports.get(spec);
+    if (importMapping) {
+      if (!isURL(importMapping)) {
+        importMapping = pathToFileURL(importMapping, this.baseURL.pathname);
+        this.imports.set(spec, importMapping);
+      }
+      return importMapping;
+    }
+
+    return specifier;
+  }
+}
+
+function processImportMap(importMap, baseURL, raw) {
+  // validation and normalization
+  // validate basic structure
+  // sort the scopes by specificity
+  if (typeof raw.imports !== 'object') {
+    throw new TypeError('import map top level key "imports" is required');
+  }
+  if (typeof raw.scopes !== 'object') {
+    throw new TypeError('import map top level key "scopes" is required');
+  }
+
+  // Normalize imports
+  for (let [specifier, mapping] of ObjectEntries(raw.imports)) {
+    if (!specifier || typeof specifier !== 'string') {
+      throw new TypeError('import map module specifier keys must be non-empty strings');
+    }
+    if (!mapping || typeof mapping !== 'string') {
+      throw new TypeError('import map module specifier values must be non-empty strings');
+    }
+    if (specifier.endsWith('/') && !mapping.endsWith('/')) {
+      throw new Error('import map module specifier values for keys ending with / must also end with /');
+    }
+
+    importMap.imports.set(specifier, mapping);
+  }
+
+  // Normalize scopes
+  for (let [scope, _scopeMap] of ObjectEntries(raw.scopes)) {
+    if (!scope || typeof scope !== 'string') {
+      throw new TypeError('import map scopes keys must be non-empty strings');
+    }
+    if (!_scopeMap || typeof _scopeMap !== 'object') {
+      throw new TypeError(`import map scope values must be plain objects (${scope} is ${typeof _scopeMap})`);
+    }
+
+    // Normalize scope
+    scope = pathToFileURL(scope, baseURL.pathname);
+
+    const scopeMap = new Map();
```

Comment:
```suggestion
    const scopeMap = new SafeMap();
```

Reply:
Oh oops, I thought I had gotten all of those. Thanks!

Reply:
I think it was something to do with the linting job. I see a lot of new lint results I didn't see before. Will be fixing those next including this.